neonvm/runner/main.go

```diff
@@ -532,10 +544,198 @@ func main() {
 		qemuCmd = append(qemuCmd, "-incoming", fmt.Sprintf("tcp:0:%d", vmv1.MigrationPort))
 	}
 
-	go terminateQemuOnSigterm(vmSpec.QMP)
-	if err := execFg(QEMU_BIN, qemuCmd...); err != nil {
+	isSystemd, err := isSystemdCgroup()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var cgroupPath string
+	cgroupName := fmt.Sprintf("%s-vm-runner", vmStatus.PodName)
+	if isSystemd {
+		cgroupPath = path.Join(cgroupSystemdBase, cgroupName)
+	} else {
+		cgroupPath = path.Join(cgroupPlainBase, cgroupName)
+	}
+
+	if err := createCgroup(cgroupPath); err != nil {
+		log.Fatal(err)
+	}
+
+	if err := setCgroupLimit(qemuCPUs.minFraction, cgroupPath); err != nil {
 		log.Fatal(err)
 	}
+
+	go terminateQemuOnSigterm(vmSpec.QMP)
+	go listenForCPUChanges(vmSpec.RunnerPort, cgroupPath)
+
+	qemuString := shellescape.QuoteCommand(append([]string{QEMU_BIN}, qemuCmd...))
+	cgropedQemu := fmt.Sprintf("echo $$ >> %s && %s", path.Join(cgroupPath, "cgroup.procs"), qemuString)
+	// start a new shell, cgroup it and then start qemu. Qemu will get cgroup from the parent sh while runner is not cgrouped
+	if err := execFg("sh", "-c", cgropedQemu); err != nil {
+		log.Println(err)
+	}
+
+	log.Println("cleaning up cgroup...")
+	err = os.Remove(cgroupPath)
+	if err != nil {
+		log.Println(err)
+	}
+	log.Println("exiting")
+}
+
+func handleCPUChange(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		log.Printf("unexpected method: %s\n", r.Method)
+		w.WriteHeader(500)
+		return
+	}
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("could not read body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	parsed := neonApi.VCPUChange{}
+	err = json.Unmarshal(body, &parsed)
+	if err != nil {
+		log.Printf("could not parse body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	cgroupPath := r.Context().Value(cgroupContextKey{})
+	val, ok := cgroupPath.(string)
+	if !ok {
+		log.Println("no path in context")
+		w.WriteHeader(500)
+		return
+	}
+
+	// update cgroup
+	fraction := getCPUFraction(parsed.VCPUs)
+	log.Printf("got CPU update %v (%v)\n", parsed.VCPUs, fraction)
+	err = setCgroupLimit(fraction, val)
+	if err != nil {
+		log.Printf("could not read body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	w.WriteHeader(200)
+}
+
+type cgroupContextKey struct{}
+
+func listenForCPUChanges(port int32, cgroupPath string) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/cpu_change", handleCPUChange)
+	baseCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	baseCtx = context.WithValue(baseCtx, cgroupContextKey{}, cgroupPath)
```

Comment:
we should do something to clean up the lifecycle of the server: 
- there's no way for the context to be canceled before the server exits
- there's no way to coordinate the server's shutdown with the process's shutdown (e.g. the process will exit before the service shuts down).